### ISSUE TITLE
bug: RxUI.EnsureInitialized called after DI container locked

### ISSUE THREAD

**Issue by weitzhandler** [reactiveui/splat#378]
**Describe the bug**
MSDI provides IoC in two stages, registration and resolution, once the provider has been built, services can no longer be added.
The problem is that RxUI does attempt to add new services after building of container.
TBH, wasn't sure where to post this issue on Rx or Splat.

**Steps To Reproduce**
- Install `Splat.Microsoft.Extensions.DependencyExtensions`
- Install RxUI
setup your app as follows:
```c#
void Init()
{
  //var location = Assembly.GetExecutingAssembly().Location;
  //var configFile = ExtractResource("Aqounts.appsettings.json", location);

  var host =
    Host
    .CreateDefaultBuilder()
    .ConfigureServices(services =>
    {
      services.UseMicrosoftDependencyResolver();
      // Configure local services and access the host configuration
      ConfigureServices(services);
    })
    .ConfigureLogging(loggingBuilder =>
    {
      // remove loggers incompatible with UWP
      {
        var eventLoggers = loggingBuilder.Services
          .Where(l => l.ImplementationType == typeof(EventLogLoggerProvider))
          .ToList();

        foreach (var el in eventLoggers)
          loggingBuilder.Services.Remove(el);
      }

      loggingBuilder.AddSplat();
    })
    .UseEnvironment(Environments.Development)
    .Build();

  //Save our service provider so we can use it later.
  Container = host.Services;
  Container.UseMicrosoftDependencyResolver();
}
```

In one of the following scenes, execute an observable which internally trigger one of the static constructors that attempts to register services.

**Solutions**
1. Silently ignore the new registrations
2. Ensure RxUI runs all service registrations when `UseMicrosoftDependencyResolver(IServiceCollection)` is called, and only if not `HasRegistration(contract)` (related: #372) - preferred and proper solution.

**Comment by RLittlesII** [comment]
@weitzhandler What is the behavior that is driving this bug?  Is there an error message, or does ReactiveUI just fail to register it's specific `IWantsToRegister`?  With all our other adapters, once we `UseDependencyResolver` generally that is the close of business.  The `update` in the `AutofacDependencyResolver` gets around the issue of container immutability.  I am not sure you can silently ignore the registrations if they are platform registrations i.e. registrations ReactiveUI needs to function.

I added [ReactiveUI.Splat.Tests](https://github.com/reactiveui/ReactiveUI/blob/master/src/ReactiveUI.Splat.Tests/SplatAdapterTests.cs) to ReactiveUI's repository to verify ReactiveUI registers it's bits.  You can start with creating the failing tests and then go from there?

I would think you would need to rebuild the container during ReactiveUI registration.

**Comment by weitzhandler** [comment]
Oh, hi @RLittlesII and thanks for your response!

It fails because it gets to this check:

https://github.com/reactiveui/splat/blob/e48eec3f17a32f13be853dae5fc8406565fdae92/src/Splat.Microsoft.Extensions.DependencyInjection/MicrosoftDependencyResolver.cs#L111-L116

It gets called on first time we call some RxUI stuff, i.e. `WhenAny`, which triggers some registrations from one of the static constructors.

**Comment by RLittlesII** [comment]
Yeah.  I think I made a comment on the PR that throwing these exceptions was not a good idea.  That none of the other adapters do this, because ReactiveUI registration would fail.

**Comment by weitzhandler** [comment]
OK, so what's the strategy we should do on fixing it.
Silently ignore it, or make sure RxUI is initialized at registration stage.
Maybe we should have another `UseMicrosoftDependencyResolver` in the RxUI namespace that calls the Splat one, and additionally initializes RxUI?

**Comment by RLittlesII** [comment]
I believe we'll have to rebuild the container.  That's the only plausible solution for now, unless you can hold onto the component registry and defer resolution of the container instance.  I say only plausible, because the point of these adapters was to be a Splat thing.  I know the lines between Splat and ReactiveUI are blurred some, but @glennawatson and I made a concerted effort during creating these packages to clearly separate the concerns.

I am currently trying to draft up some thoughts and interfaces for https://github.com/reactiveui/ReactiveUI/issues/1737 that should make this a non issue in the future.

Out of curiosity if you called `RxApp.DefaultExceptionHandler = new DefaultExceptionHandler();` or something to force the initialization *before* you finalize the container, do you run into the same issue?

**Comment by weitzhandler** [comment]
> Out of curiosity if you called `RxApp.DefaultExceptionHandler = new DefaultExceptionHandler();` or something to force the initialization _before_ you finalize the container, do you run into the same issue?

Will try that

```c#
public static void UseMicrosoftDependencyResolver(this IServiceCollection services)
{
  var resolver = new MicrosoftDependencyResolver(services);
  Locator.SetLocator(resolver);
  services.AddSingleton<IMicrosoftDependencyResolver>(resolver);

  resolver.InitializeReactiveUI();
}
```

This solves the issue for me tho.

**Comment by weitzhandler** [comment]
> I believe we'll have to rebuild the container.

I wouldn't want to go with this option, that was also my response to your comment. We have to give the option for the container to be managed externally. For instance, in my scenario it's Microsoft's [`GenericHost`](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/generic-host?view=aspnetcore-2.2#configureappconfiguration). To me Splat is there just to cater for RxUI, I use MS extensions for DI and also for building the container.

> Out of curiosity if you called `RxApp.DefaultExceptionHandler = new DefaultExceptionHandler();` 
or something to force the initialization _before_ you finalize the container, do you run into the same issue?

I'll try that

**Comment by RLittlesII** [comment]
I believe that is the entire intent of these adapters.  You just need the container to be mutable until ReactiveUI has initialized.  This is the problem we are solving .

**Comment by weitzhandler** [comment]
> Out of curiosity if you called `RxApp.DefaultExceptionHandler = new DefaultExceptionHandler();` 
or something to force the initialization _before_ you finalize the container, do you run into the same issue?

Here's what I do in my app that solves the problem:

```c#
public static void UseMicrosoftDependencyResolver(this IServiceCollection services)
{
  var resolver = new MicrosoftDependencyResolver(services);
  Locator.SetLocator(resolver);
  services.AddSingleton<IMicrosoftDependencyResolver>(resolver);

  /* ↓↓↓ */
  resolver.InitializeReactiveUI();
}
```

**Comment by weitzhandler** [comment]
> You just need the container to be mutable until ReactiveUI has initialized.

So in case of an externally managed DI container, `IntializeReactiveUI` should be called before locking it down.

**Comment by RLittlesII** [comment]
I believe that works.  Missed your above comment https://github.com/reactiveui/splat/issues/378#issuecomment-518468483

Just verify it with the tests I mentioned eariler and it should be good.

**Comment by weitzhandler** [comment]
So no need for any other implementation, just add to the docs that `InitializeReactiveUI` has to be called explicitly if we don't want to rebuild the container, right?

**Comment by glennawatson** [comment]
It's already documented https://reactiveui.net/docs/handbook/dependency-inversion/custom-dependency-inversion#set-the-locator.current-to-your-implementation

**Comment by weitzhandler** [comment]
Ok great then.
Gonna close this issue.
Sorry for the hassle.

**Comment by weitzhandler** [comment]
I now realize what's the problem, in my original implementation, it was the same instance of `IDependencyReosolver`, now that it's a different one, for some reason, it all gets called again when accessing the RxUI features such as `WhenAny`.
It fails on attmept to register `ICreatesObservableForProperty`. When I used the same instance of the resolver that didn't happen.
The resolver I used in my code before, instead of creating a new `MicrosoftDependencyResolver` (https://github.com/reactiveui/splat/blob/master/src/Splat.Microsoft.Extensions.DependencyInjection/SplatMicrosoftExtensions.cs#L31), it reused the same resolver, but made it immutable. We actually talked about it here: https://github.com/reactiveui/splat/issues/362, I just didn't realize this will be of an issue.
How about we change the mixins, so that the [second one](https://github.com/ReactiveUI/splat/blob/master/src/Splat.Microsoft.Extensions.DependencyInjection/SplatMicrosoftExtensions.cs#L30), doesn't create a new `MicrosoftDependencyResolver` but instead updates the existing one with an `IServiceProvider` and makes it immutable?
